### PR TITLE
fix: Increase precision of iso8601 date formatter

### LINKED ISSUE #857: Breadcrumb lose precision when serialize
_Platform:_
- [x]  iOS
- [ ]  tvOS
- [ ]  MacOS
- [ ]  watchOS

_Swift:_
- [x]  Yes -> If yes, which version?
- [ ]  No

_sentry-cocoa installed with:_
- [ ] Swift Package Manager
- [x] CocoaPods
- [ ] Carthage
- [ ] Manually

Version of sentry-cocoa:
**6.0.9**

---
I have following issue:

timestamp in `SentryBreadcrumb` is serizlized with 
```objc
[serializedData setValue:[self.timestamp sentry_toIso8601String] forKey:@"timestamp"];
```
and the date formatter it did not specify millisecond in the template
```objc
[isoFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss'Z'"];
```

**Steps to reproduce:**
- Send an event with breadcrumb, and view it in sentry website

**Actual result:**
- show breadcrumb with second precision

**Expected result:**
- show breadcrumb with millisecond precision

### PR DESCRIPTION
**Pull request by ainopara**
## :scroll: Description

Increase precision of iso8601 date formatter to millisecond

## :bulb: Motivation and Context

Fixes #857

## :green_heart: How did you test it?

Add more test cases.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the CHANGELOG
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

**Review by bruno-garcia — APPROVED** [review]
Great addition!

LGTM

File: Tests/SentryTests/SentryTests.m [hunk 1]
```diff
@@ -241,11 +241,27 @@ - (void)testLevelOrder
...
-            [NSDate sentry_fromIso8601String:[date sentry_toIso8601String]] timeIntervalSince1970],
-        (NSInteger)[date timeIntervalSince1970]);
+        [[NSDate sentry_fromIso8601String:[date sentry_toIso8601String]] timeIntervalSince1970],
+        timeInterval
+    );
+}
+
+- (void)testDateCategoryPrecision
+{
+    NSDate *date1 = [NSDate dateWithTimeIntervalSinceReferenceDate:0.1234];
```
Comment (philipphofmann) [inline Tests/SentryTests/SentryTests.m:254]: Very good edge case tests, thanks 👏

File: Sources/Sentry/NSDate+SentryExtras.m [hunk 2]
```diff
@@ -18,14 +18,33 @@ + (NSDateFormatter *)getIso8601Formatter
...
+
+    return isoFormatter;
+}
+
 + (NSDate *)sentry_fromIso8601String:(NSString *)string
 {
-    return [[self.class getIso8601Formatter] dateFromString:string];
+    NSDate *date = [[self.class getIso8601FormatterWithMillisecondPrecision] dateFromString:string];
+    if (nil == date) {
+        return [[self.class getIso8601Formatter] dateFromString:string];
```
Comment (philipphofmann) [inline Sources/Sentry/NSDate+SentryExtras.m:39]: `l`: Can you please add a comment here, that we do this to be backward compatible.

File: Sources/Sentry/NSDate+SentryExtras.m [hunk 3]
```diff
@@ -18,14 +18,33 @@ + (NSDateFormatter *)getIso8601Formatter
...
+    });
+
+    return isoFormatter;
+}
+
 + (NSDate *)sentry_fromIso8601String:(NSString *)string
 {
-    return [[self.class getIso8601Formatter] dateFromString:string];
+    NSDate *date = [[self.class getIso8601FormatterWithMillisecondPrecision] dateFromString:string];
+    if (nil == date) {
```
Comment (philipphofmann) [inline Sources/Sentry/NSDate+SentryExtras.m:38]: `h`: Can we please add a test for date formatted without millis, for example `2020-02-27T11:35:26Z`

**Review by philipphofmann — COMMENTED** [review]
Thanks a lot, @ainopara for this PR. Can you please address the two comments I made and please call `make format` on your console to format the code properly. Our Github Action can't commit the formatted code to your fork, unfortunately. After that, we should be good to merge 👍

File: Sources/Sentry/NSDate+SentryExtras.m (hunk 2 above)
Comment (ainopara) [reply Sources/Sentry/NSDate+SentryExtras.m:40]: done

File: Sources/Sentry/NSDate+SentryExtras.m (hunk 3 above)
Comment (ainopara) [reply Sources/Sentry/NSDate+SentryExtras.m:38]: done

**Comment by philipphofmann** [comment]
Thanks @ainopara 👏

**Review by philipphofmann — APPROVED** [review]
LGTM
